keps/sig-node/5307-container-restart-policy/README.md

```diff
@@ -0,0 +1,1159 @@
+  ContainerRestartPolicyOnFailure ContainerRestartPolicyOnFailure = "OnFailure"
+)
+
+type Container struct {
+  // Omitting irrelevant fields...
+  // RestartPolicy must be specified if RestartPolicyRules is specified.
+  RestartPolicy *ContainerRestartPolicy
+
+  // Represents a list of rules to be checked to determine if the
+  // container should be restarted on exit. The rules are evaluated in
+  // order. Once a rule matches a container exit condition, the remaining
+  // rules are ignored. If no rule matches the container exit condition,
+  // the Pod-level restart policy determines the whether the container
+  // is restarted or not. At most 20 rules are allowed.
+  RestartPolicyRules []ContainerRestartRule
+}
+
+// ContainerRestartRule describes how a container exit is handled.
+type ContainerRestartRule struct {
+  // Specifies the action taken on a container exit if the requirements
+  // are satisfied. The only possible value is "Restart" to restart the
+  // container.
+  // +required
+  Action ContainerRestartRuleAction
+
+  // Represents the exit codes to check on container exits. The oneOf
+  // field must be provided.
+  // +optional
+  // +oneOf=when
+  ExitCodes *ContainerRestartRuleOnExitCodes
+
+  // Other conditions in the future:
+  // OOMKill *ContainerRestartRuleConditionOOMKill
+  // RestartTimes *ContainerRestartRuleConditionRestartTimes
+  // Exit *ContainerRestartRuleConditionExit
+}
+
+type ContainerRestartRuleAction string
+
+const (
+  ContainerRestartRuleActionRestart ContainerRestartRuleAction = "Restart"
+
+  // Future actions: "Complete", "TerminatePod", "RestartPod".
+)
+
+// ContainerRestartRuleOnExitCodes describes the condition
+// for handling an exited container based on its exit codes.
+type ContainerRestartRuleOnExitCodes struct {
+  // Represents the relationship between the container exit code(s) and the
+	// specified values. Possible values are:
+	//
+	// - In: the requirement is satisfied if the container exit code is in the 
+  //   set of specified values.
+	// - NotIn: the requirement is satisfied if the container exit code is 
+  //   not in the set of specified values.
+  Operator ContainerRestartRuleOnExitCodesOperator
+
+  // Specifies the set of values to check for container exit codes.
+  // At most 255 elements are allowed.
+  Values []int32
+}
+
+type ContainerRestartRuleOnExitCodesOperator string
+
+const (
+  ContainerRestartRuleOnExitCodesOpIn ContainerRestartRuleOnExitCodesOperator = "In"
+  ContainerRestartRuleOnExitCodesOpNotIn ContainerRestartRuleOnExitCodesOperator = "NotIn"
+)
+```
+
+To specify a container to only restart with an exit code of 42, it
+can be specified as following in a Pod manifest:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-pod
+spec:
+  containers:
+  - name: my-container
+    image: nginx:latest
+    # restartPolicy must be specified to specify restartPolicyRules
+    restartPolicy: Never
```

Comment:
This doesn't read great :( But I understand that we can't introduce a Custom rule here.

Reply:
yes, agree, but we discussed it at length with no better option to express this without breaking existing customers